lib/Target/SubstraitPB/Import.cpp

```diff
@@ -153,6 +184,154 @@ static mlir::FailureOr<mlir::Type> importType(MLIRContext *context,
+  mlir::FailureOr<RelOpInterface> inputOp = importRel(builder, inputRel);
+  if (failed(inputOp))
+    return failure();
+  Value inputVal = inputOp.value()->getResult(0);
+
+  // Import measures if any.
+  auto measuresRegion = std::make_unique<Region>();
+  if (aggregateRel.measures_size() > 0) {
+    Block *measuresBlock = &measuresRegion->emplaceBlock();
+    measuresBlock->addArgument(inputVal.getType(), loc);
+
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointToStart(measuresBlock);
+    SmallVector<Value> measuresValues;
+    measuresValues.reserve(aggregateRel.measures_size());
+    for (const Measure &measure : aggregateRel.measures()) {
+      const AggregateFunction &aggrFunc = measure.measure();
+
+      // Import measure as `CallOp`.
+      FailureOr<CallOp> callOp = importAggregateFunction(builder, aggrFunc);
+      if (failed(callOp))
+        return failure();
+
+      measuresValues.push_back(callOp.value().getResult());
+    }
+
+    builder.create<YieldOp>(measuresValues);
+  }
+
+  // Import groupings if any.
+  auto groupingsRegion = std::make_unique<Region>();
+  SmallVector<Attribute> groupingSetsAttrs;
+  if (aggregateRel.groupings_size() > 0) {
+    Block *groupingsBlock = &groupingsRegion->emplaceBlock();
+    groupingsBlock->addArgument(inputVal.getType(), loc);
+
+    // Grouping expressions, i.e., values yielded from `groupings`.
+    SmallVector<Value> groupingExprValues;
+    groupingSetsAttrs.reserve(aggregateRel.groupings_size());
+
+    // Ops that produce unique grouping expressions. In the protobuf messages,
+    // each grouping set repeats the grouping expressions whereas the
+    // `AggregateOp` yields unique grouping expressions from the `groupings`
+    // region and represents the grouping sets as references to those.
+    llvm::SmallDenseMap<Operation *, int64_t, 16, SimpleOperationInfo>
+        groupingExprOps;
+
+    // Import one grouping set at a time.
+    for (const Grouping &grouping : aggregateRel.groupings()) {
+      // Collect references of grouping expressions for this grouping set.
+      SmallVector<int64_t> expressionRefs;
+      expressionRefs.reserve(grouping.grouping_expressions_size());
+      for (const Expression &expression : grouping.grouping_expressions()) {
+        // Import expression message into `groupings` region.
+        OpBuilder::InsertionGuard guard(builder);
+        builder.setInsertionPointToStart(groupingsBlock);
+        FailureOr<ExpressionOpInterface> exprOp =
+            importExpression(builder, expression);
+        if (failed(exprOp))
+          return failure();
+
+        // Create or look-up reference.
+        auto [it, hasInserted] = groupingExprOps.try_emplace(exprOp.value());
+
+        // If it's a new expression, assign new reference.
+        if (hasInserted) {
+          it->second = groupingExprOps.size() - 1;
+          groupingExprValues.emplace_back(exprOp.value()->getResult(0));
```

Comment:
And these will always have a single result?

Reply:
Yes, but currently only by convention. This is bothering for some time but I haven't gotten around to trying to improve things again. Ideally, I would like to attach the `OpTrait::OneResult` to the `RelOpInterface` and the `ExpressionOpInterface` but I didn't immediate manage to do that (and I am not sure if we can attach a trait to an *interface*). The best alternative I came up with is to add a `getResult()` method to the interface that simply does `return getResult(0)`.